### ISSUE TITLE
ENH: Keep netCDF files open

### ISSUE THREAD

**Issue by jhamman** [UW-Hydro/VIC#380]
We have a bunch of functions in the `image` driver that operate on netCDF files, they all look something like this:

``` C
int
get_x_from_nc_file(char   *nc_name,
                   char   *var_name,
                   size_t *x)
{

    ...

    // open the netcdf file
    status = nc_open(nc_name, NC_NOWRITE, &nc_id);
    if (status != NC_NOERR) {
        log_err("Error opening %s", nc_name);
    }

    /* get NetCDF variable */
    status = nc_inq_varid(nc_id, var_name, &var_id);
    if (status != NC_NOERR) {
        log_err("Error getting variable id for %s in %s", var_name, nc_name);
    }

...

    // close the netcdf file
    status = nc_close(nc_id);
    if (status != NC_NOERR) {
        log_err("Error closing %s", nc_name);
    }

    return status;
}
```

We end up redundantly opening and closing the parameter, forcings, and domain files.  I think @wietsefranssen has already reported that keeping the forcing file open speeds up the image driver considerably, the next step would be to extend that to all netcdf files in the image driver, removing the open/close lines from many of these functions.  We will just need to create structures to store the file pointers (e.g. `nc_id`) and functions to open/close them at the beginning and end of a simulation. 

@bartnijssen, am I missing anything here?

**Comment by bartnijssen** [comment]
This was just implemented that way for expediency sake when doing the first implementation. We *definitely* want to keep the files open during the run, since the reopening of the files brings a lot of overhead for netcdf files. Similarly we'll want to keep the output files open for as long as we can.

Note though that you cannot necessarily keep the files open for the duration of the simulation since the files maybe organized by year. So you need to store the start and end time of a file and figure out whether you need to open a new file or can reuse the one that is already open.

**Comment by wietsefranssen** [comment]
A while ago I did a little test with keeping the files open and it indeed increased the running time a lot.
I am looking forward to the changes!

**Comment by bartnijssen** [comment]
Me too!

**Comment by bartnijssen** [comment]
Bumping this to 5.1. I think we already have a few ways of how we may want to implement this. Not sure how separate this is from #538 (seems mostly a duplicate to me, so we could potentially get rid of at least one of them).

**Comment by bartnijssen** [comment]
Rather than using `nc_open` and `nc_close` every time we read a netcdf entry (field, dimension, variable, attribute, etc), we should `nc_close` only once (take it out of the read functions) and check whether the file is open before we call `nc_open`. For example, you could expand the use of `nc_file_struct` (https://github.com/UW-Hydro/VIC/blob/master/vic/drivers/shared_image/include/vic_driver_shared_image.h#L104-L142) and pass or perhaps something simpler instead of the filename to the functions that read fields. For example:

``` C
/******************************************************************************
 * @brief    Read double precision netCDF field from file.
 *****************************************************************************/
int
get_nc_field_double(char   *nc_name,
                    char   *var_name,
                    size_t *start,
                    size_t *count,
                    double *var)
{
    int nc_id;
    int status;
    int var_id;

    // open the netcdf file
    status = nc_open(nc_name, NC_NOWRITE, &nc_id);
    check_nc_status(status, "Error opening %s", nc_name);
```

could become

/******************************************************************************
 * @brief    Read double precision netCDF field from file.
 *****************************************************************************/
int
get_nc_field_double(nc_file_struct   *nc_file,
                    char   *var_name,
                    size_t *start,
                    size_t *count,
                    double *var)
{
    int nc_id;
    int status;
    int var_id;

    // open the netcdf file
    if (nc_file->open == FALSE) {
        status = nc_open(nc_file->name, NC_NOWRITE, &nc_id);
        check_nc_status(status, "Error opening %s", nc_file->name);
        nc_file->open = TRUE
    }
```

When you call `nc_close()` you would then also do `ncfile->open = FALSE`.

As said you will have to think a bit on how to implement this, but it will likely increase the initialization by a lot, because we open and close the files many times, which with netcdf invokes a lot of overhead. As said at the start of this issue: the current implementation was simply for expediency sake while implementing netcdf in VIC. It was never end as the production solution, so I'll leave it up to @yixinmao to implement the production version.

**Comment by bartnijssen** [comment]
Closed per #684
